fdb-extensions/src/main/java/com/apple/foundationdb/async/MoreAsyncUtil.java

```diff
@@ -857,6 +857,33 @@ public static void closeIterator(@Nonnull Iterator<?> iterator) {
         }
     }
 
+    /**
+     * Compose a handler bi-function to the result of a future. Unlike the
+     * {@link AsyncUtil#composeHandle(CompletableFuture, BiFunction)}, which completes exceptionally only when the
+     * <code>handler</code> completes exceptionally, it completes exceptionally even if the supplied action itself
+     * (<code>future</code>) encounters an exception.
+     * @param future future to compose the handler onto
+     * @param handler handler bi-function to compose onto the passed future
+     * @param exceptionMapper function for mapping the underlying exception to a {@link RuntimeException}
+     * @param <V> return type of original future
+     * @param <T> return type of final future
+     * @return future with same completion properties as the future returned by the handler
+     * @see AsyncUtil#composeHandle(CompletableFuture, BiFunction)
+     */
+    public static <V, T> CompletableFuture<T> composeWhenComplete(
+            @Nonnull CompletableFuture<V> future,
+            @Nonnull BiFunction<V,Throwable,? extends CompletableFuture<T>> handler,
+            @Nullable Function<Throwable,RuntimeException> exceptionMapper) {
+        return AsyncUtil.composeHandle(future, (result, ex) -> handler.apply(result, ex)
+                .thenApply(handlerResult -> {
```

Comment:
Because this calls `.thenApply`, if running the handler returns an exceptional future, then we get the error back from the handler rather than original function. The semantics of `.whenComplete` are that it returns back the error from `future` if it errored, or the error from the `handler` if it errored, or the original result if both completed successfully. So that would mean using `.handle` here, and then remapping the appropriate exception (if we want to replicate the semantics).

In theory, `handler.apply` could *also* throw an error (rather than returning an exceptional future), and one might want to also catch that error to know whether to throw that one or the error from the original future.

Reply:
Fixed both.